README.md
[![Build Status](https://travis-ci.org/avkhozov/Minion-Backend-Mango.svg?branch=master)](https://travis-ci.org/avkhozov/Minion-Backend-Mango)
# NAME

Minion::Backend::Mango - Mango backend for Minion

# SYNOPSIS

    use Minion::Backend::Mango;

    my $backend = Minion::Backend::Mango->new('mongodb://127.0.0.1:27017');

# DESCRIPTION

[Minion::Backend::Mango](https://metacpan.org/pod/Minion::Backend::Mango) is a highly scalable [Mango](https://metacpan.org/pod/Mango) backend for [Minion](https://metacpan.org/pod/Minion).

# ATTRIBUTES

[Minion::Backend::Mango](https://metacpan.org/pod/Minion::Backend::Mango) inherits all attributes from [Minion::Backend](https://metacpan.org/pod/Minion::Backend) and
implements the following new ones.

## mango

    my $mango = $backend->mango;
    $backend  = $backend->mango(Mango->new);

[Mango](https://metacpan.org/pod/Mango) object used to store collections.

## jobs

    my $jobs = $backend->jobs;
    $backend = $backend->jobs(Mango::Collection->new);

[Mango::Collection](https://metacpan.org/pod/Mango::Collection) object for `jobs` collection, defaults to one based on ["prefix"](#prefix).

## notifications

    my $notifications = $backend->notifications;
    $backend          = $backend->notifications(Mango::Collection->new);

[Mango::Collection](https://metacpan.org/pod/Mango::Collection) object for `notifications` collection, defaults to one based on ["prefix"](#prefix).

## prefix

    my $prefix = $backend->prefix;
    $backend   = $backend->prefix('foo');

Prefix for collections, defaults to `minion`.

## workers

    my $workers = $backend->workers;
    $backend    = $backend->workers(Mango::Collection->new);

[Mango::Collection](https://metacpan.org/pod/Mango::Collection) object for `workers` collection, defaults to one based on ["prefix"](#prefix).

# METHODS

[Minion::Backend::Mango](https://metacpan.org/pod/Minion::Backend::Mango) inherits all methods from [Minion::Backend](https://metacpan.org/pod/Minion::Backend) and implements the following new ones.

## dequeue

    my $job_info = $backend->dequeue($worker_id, 0.5);
    my $job_info = $backend->dequeue($worker_id, 0.5, {queues => ['default']});

Wait for job, dequeue it and transition from `inactive` to `active` state or
return `undef` if queue was empty.

These options are currently available:

- queues

        queues => ['important']

    One or more queues to dequeue jobs from, defaults to `default`.

## enqueue

    my $job_id = $backend->enqueue('foo');
    my $job_id = $backend->enqueue(foo => [@args]);
    my $job_id = $backend->enqueue(foo => [@args] => {priority => 1});

Enqueue a new job with `inactive` state. These options are currently available:

- attempts

        attempts => 25

    Number of times performing this job will be attempted, defaults to `1`.

- delay

        delay => 10

    Delay job for this many seconds from now.

- priority

        priority => 5

    Job priority, defaults to `0`.

- queue

        queue => 'important'

    Queue to put job in, defaults to `default`.

## fail\_job

    my $bool = $backend->fail_job($job_id);
    my $bool = $backend->fail_job($job_id, 'Something went wrong!');

Transition from `active` to `failed` state.

## finish\_job

    my $bool = $backend->finish_job($job_id);

Transition from `active` to `finished` state.

## job\_info

    my $info = $backend->job_info($job_id);

Get information about a job or return `undef` if job does not exist.

## list\_jobs

    my $batch = $backend->list_jobs($skip, $limit);
    my $batch = $backend->list_jobs($skip, $limit, {state => 'inactive'});

Returns the same information as ["job\_info"](#job_info) but in batches.

These options are currently available:

- state

        state => 'inactive'

    List only jobs in this state.

- task

        task => 'test'

    List only jobs for this task.

## list\_workers

    my $batch = $backend->list_workers($skip, $limit);

Returns the same information as ["worker\_info"](#worker_info) but in batches.

## new

    my $backend = Minion::Backend::Mango->new('mongodb://127.0.0.1:27017');

Construct a new [Minion::Backend::Mango](https://metacpan.org/pod/Minion::Backend::Mango) object.

## register\_worker

    my $worker_id = $backend->register_worker;
    my $worker_id = $backend->register_worker($worker_id);

Register worker or send heartbeat to show that this worker is still alive.

## remove\_job

    my $bool = $backend->remove_job($job_id);

Remove `failed`, `finished` or `inactive` job from queue.

## repair

    $backend->repair;

Repair worker registry and job queue if necessary.

## reset

    $backend->reset;

Reset job queue.

## retry\_job

    my $bool = $backend->retry_job($job_id);
    my $bool = $backend->retry_job($job_id, {delay => 10});

Transition from `failed` or `finished` state back to `inactive`, already
`inactive` jobs may also be retried to change options.

These options are currently available:

- delay

        delay => 10

    Delay job for this many seconds (from now).

- priority

        priority => 5

    Job priority.

- queue

        queue => 'important'

    Queue to put job in.

## stats

    my $stats = $backend->stats;

Get statistics for jobs and workers.

## unregister\_worker

    $backend->unregister_worker($worker_id);

Unregister worker.

## worker\_info

    my $info = $backend->worker_info($worker_id);

Get information about a worker or return `undef` if worker does not exist.

# AUTHOR

Andrey Khozov <avkhozov@gmail.com>

Sebastian Riedel <sri@cpan.org>

# LICENSE

Copyright (C) 2014, Sebastian Riedel.

Copyright (C) 2015-2016, Andrey Khozov.

This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

# SEE ALSO

[Minion](https://metacpan.org/pod/Minion), [Mango](https://metacpan.org/pod/Mango), [http://mojolicio.us](http://mojolicio.us).


Changes
Revision history for Perl extension Minion-Backend-Mango

{{$NEXT}}

1.02 2017-06-06T05:20:09Z
    - Implement broadcast and receive methods to allow use worker remote control
      with Mango backend.

1.01 2017-03-29T06:17:56Z
    - Added dummy broadcast and receive methods to allow use Mango backend
      with new versions of Minion.

1.00 2016-05-03T10:32:20Z
    - Added queue options to list_jo method.
    - Added delayed_jobs field to stats.
    - Improved tests for new Minion 5 features.

0.99_50 2016-05-03T07:56:26Z
    - Fixed bug with wrong removing abandoned jobs.

0.99 2016-01-04T20:11:41Z
    - Fully compatible with Minion 4.01.
    - Fixed bug with wait timeout in dequeue.
    - Fixed bug with update jobs if worker went away.
    - Improved retry method for allow retry already inactive jobs.
    - Improved tests.

0.98 2015-10-30T10:04:49Z
    - Added support for retrying jobs with a new priority.
    - Added support for queues.
    - Added attempts option for enqueue method.
    - Improved tests.

0.97 2015-05-19T04:28:55Z
    - Improved documentation.
    - Fixed versions in cpanfile: Minion (1.15), Mango (1.17).

0.96 2015-05-18T17:58:06Z
    - Added support for retrying jobs with delays.
    - Improved job_info method for compatibility with Minion. Renamed error
      to result field.

0.95 2015-03-13T07:07:47Z
    - Added notified field to worker_info output.

0.94 2015-03-12T16:07:19Z
    - Added support for monitoring workers with heartbeats.
    - Set limit for capped collection size to 128.
    - Improved performance by following only new jobs at the start.

0.93 2015-03-07T08:51:04Z
    - Added indexes for jobs collection.

0.92 2015-01-17T07:24:14Z
    - Improved timeout management.
    - Capped collection for notifications limited only by size, not max number
      of documents.

0.91 2015-01-07T17:59:19Z
    - First authorized release.

0.90 2015-01-06T17:07:25Z
    - First release.


lib/Minion/Backend/Mango.pm
package Minion::Backend::Mango;
use Mojo::Base 'Minion::Backend';

our $VERSION = '1.01';

use Mango;
use Mango::BSON qw(bson_oid bson_time bson_doc);
use Sys::Hostname 'hostname';
use Time::HiRes 'time';

has 'mango';
has jobs          => sub { $_[0]->mango->db->collection($_[0]->prefix . '.jobs') };
has notifications => sub { $_[0]->mango->db->collection($_[0]->prefix . '.notifications') };
has prefix        => 'minion';
has workers       => sub { $_[0]->mango->db->collection($_[0]->prefix . '.workers') };

sub broadcast {
  my ($self, $command, $args, $ids) = (shift, shift, shift || [], shift || []);

  my $opts = @$ids ? {_id => {'$in' => $ids}} : {};
  return !!$self->workers->update($opts, {'$push' => {inbox => [$command, @$args]}}, {multi => 1})->{n};
}

sub dequeue {
  my ($self, $wid, $wait, $options) = @_;

  $self->_notifications;

  my $end = time + $wait;
  do {
    if (my $job = $self->_try($wid, $options)) { return $self->_job_info($job); }
    $self->_await;
  } while time < $end;

  return $self->_job_info($self->_try($wid, $options));
}

sub enqueue {
  my ($self, $task) = (shift, shift);
  my $args    = shift // [];
  my $options = shift // {};

  $self->_notifications;

  my $doc = {
    args     => $args,
    attempts => $options->{attempts} // 1,
    created  => bson_time,
    delayed  => bson_time($options->{delay} ? (time + $options->{delay}) * 1000 : 1),
    priority => $options->{priority} // 0,
    queue    => $options->{queue}    // 'default',
    retries  => 0,
    state    => 'inactive',
    task     => $task
  };

  my $oid = $self->jobs->insert($doc);
  $self->notifications->insert({c => 'created'});
  return $oid;
}

sub fail_job { shift->_update(1, @_) }

sub finish_job { shift->_update(0, @_) }

sub job_info {
  $_[0]->_job_info($_[0]->jobs->find_one(bson_oid($_[1])));
}

sub list_jobs {
  my ($self, $skip, $limit, $options) = @_;

  my $cursor = $self->jobs->find({state => {'$exists' => \1}});
  for (qw/state task queue/) {
    $cursor->query->{$_} = $options->{$_} if $options->{$_};
  }
  $cursor->sort({_id => -1})->skip($skip)->limit($limit);

  return [map { $self->_job_info($_) } @{$cursor->all}];
}

sub list_workers {
  my ($self, $skip, $limit) = @_;
  my $cursor = $self->workers->find({pid => {'$exists' => \1}});
  $cursor->sort({_id => -1})->skip($skip)->limit($limit);
  return [map { $self->_worker_info($_) } @{$cursor->all}];
}

sub new { shift->SUPER::new(mango => Mango->new(@_)) }

sub receive {
  my ($self, $id) = @_;

  my $doc = $self->workers->find_and_modify({query => {_id => $id}, update => {'$set' => {inbox => []}}});
  return $doc->{inbox} // [];
}

sub register_worker {
  my ($self, $id, $options) = (shift, shift, shift || {});

  my $status = $options->{status} // {};
  return $id
    if $id
    && $self->workers->find_and_modify(
    {query => {_id => $id}, update => {'$set' => {notified => bson_time, status => $status}}});

  $self->jobs->ensure_index(bson_doc(state => 1, delayed => 1, task => 1, queue => 1));
  return $self->workers->insert(
    {host => hostname, pid => $$, started => bson_time, notified => bson_time, status => $status});
}

sub remove_job {
  my ($self, $oid) = @_;
  my $doc = {_id => $oid, state => {'$in' => [qw(failed finished inactive)]}};
  return !!$self->jobs->remove($doc)->{n};
}

sub repair {
  my $self   = shift;
  my $minion = $self->minion;

  # Check worker registry
  my $workers = $self->workers;
  $workers->remove({notified => {'$lt' => bson_time((time - $minion->missing_after) * 1000)}});

  # Abandoned jobs
  my $jobs = $self->jobs;
  my $cursor = $jobs->find({state => 'active'}, {_id => 1, retries => 1, worker => 1});
  while (my $job = $cursor->next) {
    $self->fail_job($job->{_id}, $job->{retries}, 'Worker went away')
      unless $workers->find_one($job->{worker});
  }

  # Old jobs
  $jobs->remove(
    {state => 'finished', finished => {'$lt' => bson_time((time - $minion->remove_after) * 1000)}});
}

sub reset {
  my $self = shift;

  $_->options && $_->drop for $self->workers, $self->jobs, $self->notifications;
  delete $self->{capped};
}

sub retry_job {
  my ($self, $oid, $retries) = (shift, shift, shift);
  my $options = shift // {};

  my $query = {_id => $oid, retries => $retries, state => {'$in' => [qw(failed finished inactive)]}};
  my $update = {
    '$inc' => {retries => 1},
    '$set' => {
      retried => bson_time,
      state   => 'inactive',
      delayed => bson_time($options->{delay} ? (time + $options->{delay}) * 1000 : 1),
      (defined $options->{priority} ? (priority => $options->{priority}) : ()),
      (defined $options->{queue}    ? (queue    => $options->{queue})    : ())
    }
  };

  return !!$self->jobs->update($query, $update)->{n};
}

sub stats {
  my $self = shift;

  my $jobs    = $self->jobs;
  my $active  = @{$jobs->find({state => 'active'})->distinct('worker')};
  my $workers = $self->workers;
  my $all     = $workers->find->count;
  my $stats   = {active_workers => $active, inactive_workers => $all - $active};
  $stats->{"${_}_jobs"} = $jobs->find({state => $_})->count for qw(active failed finished inactive);
  $stats->{delayed_jobs} = $jobs->find({state => 'inactive', delayed => {'$gt' => bson_time}})->count;
  return $stats;
}

sub unregister_worker { shift->workers->remove(shift) }

sub worker_info { $_[0]->_worker_info($_[0]->workers->find_one($_[1])) }

sub _await {
  my $self = shift;

  my $last = $self->{last} //= bson_oid;
  my $cursor
    = $self->notifications->find({_id => {'$gt' => $last}, c => 'created'})->tailable(1)->await_data(1);
  return undef unless my $doc = $cursor->next || $cursor->next;
  $self->{last} = $doc->{_id};
  return 1;
}

sub _job_info {
  my $self = shift;

  return undef unless my $job = shift;
  return {
    args     => $job->{args},
    attempts => $job->{attempts},
    created  => $job->{created} ? $job->{created}->to_epoch : undef,
    delayed  => $job->{delayed} ? $job->{delayed}->to_epoch : undef,
    finished => $job->{finished} ? $job->{finished}->to_epoch : undef,
    id       => $job->{_id},
    priority => $job->{priority},
    queue    => $job->{queue},
    result   => $job->{result},
    retried  => $job->{retried} ? $job->{retried}->to_epoch : undef,
    retries => $job->{retries} // 0,
    started => $job->{started} ? $job->{started}->to_epoch : undef,
    state   => $job->{state},
    task    => $job->{task}
  };
}

sub _notifications {
  my $self = shift;

  # We can only await data if there's a document in the collection
  $self->{capped} ? return : $self->{capped}++;
  my $notifications = $self->notifications;
  return if $notifications->options;
  $notifications->create({capped => \1, size => 1048576, max => 128});
  $notifications->insert({});
}

sub _try {
  my ($self, $wid, $options) = @_;

  my $doc = {
    query => bson_doc(
      state   => 'inactive',
      delayed => {'$lt' => bson_time},
      task    => {'$in' => [keys %{$self->minion->tasks}]},
      queue   => {'$in' => $options->{queues} || ['default']}
    ),
    fields => bson_doc(args     => 1,  retries => 1, task => 1),
    sort   => bson_doc(priority => -1, created => 1),
    update => {'$set' => {started => bson_time, state => 'active', worker => $wid}},
    new    => 1
  };

  return $self->jobs->find_and_modify($doc);
}

sub _update {
  my ($self, $fail, $oid, $retries, $result) = @_;

  my $update = {finished => bson_time, result => $result, state => $fail ? 'failed' : 'finished'};
  my $query = {_id => $oid, retries => $retries, state => 'active'};

  my $opts = {query => $query, update => {'$set' => $update}, fields => {attempts => 1}};
  return undef unless my $job = $self->jobs->find_and_modify($opts);
  return 1 if !$fail || (my $attempts = $job->{attempts}) == 1;
  return 1 if $retries >= ($attempts - 1);
  my $delay = $self->minion->backoff->($retries);
  return $self->retry_job($oid, $retries, {delay => $delay});
}

sub _worker_info {
  my $self = shift;

  return undef unless my $worker = shift;
  my $cursor = $self->jobs->find({state => 'active', worker => $worker->{_id}});
  return {
    host     => $worker->{host},
    id       => $worker->{_id},
    jobs     => [map { $_->{_id} } @{$cursor->all}],
    pid      => $worker->{pid},
    started  => $worker->{started}->to_epoch,
    notified => $worker->{notified}->to_epoch
  };
}

1;

=encoding utf8

=head1 NAME

Minion::Backend::Mango - Mango backend for Minion

=head1 SYNOPSIS

  use Minion::Backend::Mango;

  my $backend = Minion::Backend::Mango->new('mongodb://127.0.0.1:27017');

=head1 DESCRIPTION

L<Minion::Backend::Mango> is a highly scalable L<Mango> backend for L<Minion>.

=head1 ATTRIBUTES

L<Minion::Backend::Mango> inherits all attributes from L<Minion::Backend> and
implements the following new ones.

=head2 mango

  my $mango = $backend->mango;
  $backend  = $backend->mango(Mango->new);

L<Mango> object used to store collections.

=head2 jobs

  my $jobs = $backend->jobs;
  $backend = $backend->jobs(Mango::Collection->new);

L<Mango::Collection> object for C<jobs> collection, defaults to one based on L</"prefix">.

=head2 notifications

  my $notifications = $backend->notifications;
  $backend          = $backend->notifications(Mango::Collection->new);

L<Mango::Collection> object for C<notifications> collection, defaults to one based on L</"prefix">.

=head2 prefix

  my $prefix = $backend->prefix;
  $backend   = $backend->prefix('foo');

Prefix for collections, defaults to C<minion>.

=head2 workers

  my $workers = $backend->workers;
  $backend    = $backend->workers(Mango::Collection->new);

L<Mango::Collection> object for C<workers> collection, defaults to one based on L</"prefix">.

=head1 METHODS

L<Minion::Backend::Mango> inherits all methods from L<Minion::Backend> and implements the following new ones.

=head2 dequeue

  my $job_info = $backend->dequeue($worker_id, 0.5);
  my $job_info = $backend->dequeue($worker_id, 0.5, {queues => ['default']});

Wait for job, dequeue it and transition from C<inactive> to C<active> state or
return C<undef> if queue was empty.

These options are currently available:

=over 2

=item queues

  queues => ['important']

One or more queues to dequeue jobs from, defaults to C<default>.

=back

=head2 enqueue

  my $job_id = $backend->enqueue('foo');
  my $job_id = $backend->enqueue(foo => [@args]);
  my $job_id = $backend->enqueue(foo => [@args] => {priority => 1});

Enqueue a new job with C<inactive> state. These options are currently available:

=over 2

=item attempts

  attempts => 25

Number of times performing this job will be attempted, defaults to C<1>.

=item delay

  delay => 10

Delay job for this many seconds from now.

=item priority

  priority => 5

Job priority, defaults to C<0>.

=item queue

  queue => 'important'

Queue to put job in, defaults to C<default>.

=back

=head2 fail_job

  my $bool = $backend->fail_job($job_id);
  my $bool = $backend->fail_job($job_id, 'Something went wrong!');

Transition from C<active> to C<failed> state.

=head2 finish_job

  my $bool = $backend->finish_job($job_id);

Transition from C<active> to C<finished> state.

=head2 job_info

  my $info = $backend->job_info($job_id);

Get information about a job or return C<undef> if job does not exist.

=head2 list_jobs

  my $batch = $backend->list_jobs($skip, $limit);
  my $batch = $backend->list_jobs($skip, $limit, {state => 'inactive'});

Returns the same information as L</"job_info"> but in batches.

These options are currently available:

=over 2

=item queue

  queue => 'important'

List only jobs in this queue.

=item state

  state => 'inactive'

List only jobs in this state.

=item task

  task => 'test'

List only jobs for this task.

=back

=head2 list_workers

  my $batch = $backend->list_workers($skip, $limit);

Returns the same information as L</"worker_info"> but in batches.

=head2 new

  my $backend = Minion::Backend::Mango->new('mongodb://127.0.0.1:27017');

Construct a new L<Minion::Backend::Mango> object.

=head2 register_worker

  my $worker_id = $backend->register_worker;
  my $worker_id = $backend->register_worker($worker_id);

Register worker or send heartbeat to show that this worker is still alive.

=head2 remove_job

  my $bool = $backend->remove_job($job_id);

Remove C<failed>, C<finished> or C<inactive> job from queue.

=head2 repair

  $backend->repair;

Repair worker registry and job queue if necessary.

=head2 reset

  $backend->reset;

Reset job queue.

=head2 retry_job

  my $bool = $backend->retry_job($job_id);
  my $bool = $backend->retry_job($job_id, {delay => 10});

Transition from C<failed> or C<finished> state back to C<inactive>, already
C<inactive> jobs may also be retried to change options.

These options are currently available:

=over 2

=item delay

  delay => 10

Delay job for this many seconds (from now).

=item priority

  priority => 5

Job priority.

=item queue

  queue => 'important'

Queue to put job in.

=back

=head2 stats

  my $stats = $backend->stats;

Get statistics for jobs and workers.

=head2 unregister_worker

  $backend->unregister_worker($worker_id);

Unregister worker.

=head2 worker_info

  my $info = $backend->worker_info($worker_id);

Get information about a worker or return C<undef> if worker does not exist.

=head1 AUTHOR

Andrey Khozov E<lt>avkhozov@gmail.comE<gt>

Sebastian Riedel E<lt>sri@cpan.orgE<gt>

=head1 LICENSE

Copyright (C) 2014, Sebastian Riedel.

Copyright (C) 2015-2016, Andrey Khozov.

This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=head1 SEE ALSO

L<Minion>, L<Mango>, L<http://mojolicio.us>.

=cut
